### PR TITLE
Add fuzzer for `validateTx`

### PR DESCRIPTION
**Pull request by facuMH**
This PR adds a fuzzer targets `validateTx` in the TxPool. 

It will generate random values for tx':
- type (legacy, accessList, dynamicFee, blob and setCode)
- nonce
- gas
- gas price (or fee cap)
- gas tip
- value

and then for the block and pool:
- current block number
- base fee
- available gas in current block
- current revision

The objective of this test is not only to create random values to try to trigger panics in `validateTx` but also to ensure that if `validate Tx` approves a transaction, then a processor with the same state and context will also be able to process it.
To ensure this, the latter set of values will be used for both, validate options as well as configuring the evm for the processor.

### CODE REVIEW

File: evmcore/tx_validation_fuzzer_test.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+		for _, stateBalance := range []*big.Int{underCost, cost, overCost} {
+			for _, stateNonce := range []uint64{nonce - 1, nonce, nonce + 1} {
+
+				state := state.NewMockStateDB(ctxt)
+				state.EXPECT().GetBalance(from).Return(uint256.MustFromBig(stateBalance)).AnyTimes()
+				state.EXPECT().GetNonce(from).Return(stateNonce).AnyTimes()
+				stateExpectCalls(state)
+
+				// TODO: fuzz on validation options as well.
```
Comment (LuisPH3) [inline evmcore/tx_validation_fuzzer_test.go:79]: missing feature.
I think validation options should be fuzzed as well. yes.

File: evmcore/tx_validation_fuzzer_test.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+	// Seed corpus with a few valid-looking values
+	f.Add(uint8(2), uint64(1), uint64(42_000), int64(1_000), int64(500), int64(0),
+		[]byte("hi"), false,
+		int64(1_000_000_000), int64(10), uint64(50_000), int8(3),
+	)
+
+	f.Fuzz(func(t *testing.T,
+		// transaction values
+		txType uint8, nonce, gas uint64, feeCap, tip, value int64,
+		data []byte, isCreate bool,
```
Comment (LuisPH3) [inline evmcore/tx_validation_fuzzer_test.go:41]: Missing other data affecting intrinsic gas computation:
- access list
- auth list.

File: evmcore/tx_validation_fuzzer_test.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+	state.EXPECT().AddLog(any).AnyTimes()
+	state.EXPECT().AddRefund(any).AnyTimes()
+
+	state.EXPECT().SetCode(any, any).Return([]byte{}).AnyTimes()
+	state.EXPECT().SetNonce(any, any, any).AnyTimes()
+	state.EXPECT().SetState(any, any, any).Return(common.Hash{}).AnyTimes()
+
+	state.EXPECT().AddAddressToAccessList(any).AnyTimes()
+	state.EXPECT().Prepare(any, any, any, any, any, any).AnyTimes()
+	state.EXPECT().SubBalance(any, any, any).AnyTimes()
```
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go:223]: GetBalance for the sender account is called at [line75](https://github.com/0xsoniclabs/sonic/pull/196/files#diff-4f9ac9a1333876a5a8c5f9b54e39f8ec275e8fa6cc4f9791c648c7ad45377ab9R75) with the custom balance same with the nonce for the sender account

File: evmcore/tx_validation_fuzzer_test.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/holiman/uint256"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+)
+
+const (
+	testIstanbul = 0x00
```
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go]: https://github.com/0xsoniclabs/sonic/pull/196/commits/35a09d255c9e93147c408f0319cbecca5bf9327b#diff-4f9ac9a1333876a5a8c5f9b54e39f8ec275e8fa6cc4f9791c648c7ad45377ab9L21

File: evmcore/tx_validation_fuzzer_test.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+}
+
+func stateExpectCalls(state *state.MockStateDB) {
+	// expected calls to the state
+	any := gomock.Any()
+	state.EXPECT().Snapshot().AnyTimes()
+	state.EXPECT().RevertToSnapshot(any).AnyTimes()
+
+	// all accounts are unknown to a new state
+	state.EXPECT().Exist(any).Return(false).AnyTimes()
```
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go]: early exit was not triggered because eip158 was not set ([geth/core/vm/evm.go#L218](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L218)), but it should have been. So I enabled eip158 (https://github.com/0xsoniclabs/sonic/pull/196/commits/35a09d255c9e93147c408f0319cbecca5bf9327b?diff=unified&w=0#diff-4f9ac9a1333876a5a8c5f9b54e39f8ec275e8fa6cc4f9791c648c7ad45377ab9R279) and now return `true` to prevent creates or early returns.

File: evmcore/tx_validation_fuzzer_test.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+	state.EXPECT().CreateAccount(any).AnyTimes()
+	state.EXPECT().CreateContract(any).AnyTimes()
+	state.EXPECT().EndTransaction().AnyTimes()
+	state.EXPECT().TxIndex().Return(4).AnyTimes()
+
+	state.EXPECT().GetCode(any).Return([]byte{}).AnyTimes()
+	state.EXPECT().GetCodeHash(any).Return(common.Hash{}).AnyTimes()
```
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go]: https://github.com/0xsoniclabs/sonic/pull/196/commits/35a09d255c9e93147c408f0319cbecca5bf9327b?diff=unified&w=0#diff-4f9ac9a1333876a5a8c5f9b54e39f8ec275e8fa6cc4f9791c648c7ad45377ab9R248

File: evmcore/tx_validation_fuzzer_test.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+		chainId := big.NewInt(84)
+
+		tx := makeTxOfType(txType, nonce, gas, feeCap, tip, data, value, isCreate, chainId)
+		from, signedTx := signTxForTestWithChainId(t, tx, chainId)
+
+		cost := signedTx.Cost()
+		underCost := new(big.Int).Sub(cost, big.NewInt(1))
+		overCost := new(big.Int).Add(cost, big.NewInt(1))
+
+		for _, stateBalance := range []*big.Int{underCost, cost, overCost} {
```
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go:96]: https://github.com/0xsoniclabs/sonic/pull/196/commits/35a09d255c9e93147c408f0319cbecca5bf9327b?diff=unified&w=0#diff-4f9ac9a1333876a5a8c5f9b54e39f8ec275e8fa6cc4f9791c648c7ad45377ab9R91

File: evmcore/tx_validation_fuzzer_test.go (hunk 1 above)
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go]: https://github.com/0xsoniclabs/sonic/pull/196/commits/35a09d255c9e93147c408f0319cbecca5bf9327b?diff=unified&w=0#diff-4f9ac9a1333876a5a8c5f9b54e39f8ec275e8fa6cc4f9791c648c7ad45377ab9R98

File: evmcore/tx_validation_fuzzer_test.go (hunk 2 above)
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go:56]: https://github.com/0xsoniclabs/sonic/pull/196/commits/35a09d255c9e93147c408f0319cbecca5bf9327b?diff=unified&w=0#diff-4f9ac9a1333876a5a8c5f9b54e39f8ec275e8fa6cc4f9791c648c7ad45377ab9R59

File: evmcore/tx_validation_fuzzer_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (LuisPH3) [reply evmcore/tx_validation_fuzzer_test.go:96]: please rephrase, you explain twice the values but now why the 4 domains do not need to be fuzzed. 
(remember to explain why, and not what... I can see that they are the -1 and +1 values, I just don't know why to these 4 domains and not to all the domains in the fuzzer)

File: evmcore/tx_validation_fuzzer_test.go [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+	state.EXPECT().RevertToSnapshot(any).AnyTimes()
+
+	// all accounts are unknown to a new state
+	state.EXPECT().Exist(any).Return(false).AnyTimes()
+	state.EXPECT().CreateAccount(any).AnyTimes()
+	state.EXPECT().CreateContract(any).AnyTimes()
+	state.EXPECT().EndTransaction().AnyTimes()
+	state.EXPECT().TxIndex().Return(4).AnyTimes()
+
+	state.EXPECT().GetCode(any).Return([]byte{}).AnyTimes()
```
[the remark this answers is not in the document]
Comment (LuisPH3) [reply evmcore/tx_validation_fuzzer_test.go:246]: But we do test prague as well.

File: evmcore/tx_validation_fuzzer_test.go [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+	to := common.Address{0x42}
+	toPtr := &to
+	if isCreate {
+		// Set to nil
+		toPtr = nil
+	}
+
+	accessList := make([]types.AccessTuple, accessListSize)
+	for i := range accessListSize {
+		accessList[i] = types.AccessTuple{}
```
Comment (LuisPH3) [inline evmcore/tx_validation_fuzzer_test.go:160]: intrinsic cost depends on address and keys per address. they are accounted separatelly.

File: evmcore/tx_validation_fuzzer_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go:269]: yes, sorry, this is the correct reference https://github.com/0xsoniclabs/sonic/pull/196/files#diff-4f9ac9a1333876a5a8c5f9b54e39f8ec275e8fa6cc4f9791c648c7ad45377ab9R103

File: Makefile [hunk 10]
```diff
@@ -61,6 +61,21 @@ integration-coverage:
...
+fuzz-validate-cover: TEMPSEEDDIR=./evmcore/testdata/fuzz/FuzzValidateTransaction/
+fuzz-validate-cover:
+	@mkdir -p ${GOCOVERDIR} ;\
+     mkdir -p ${TEMPSEEDDIR} ;\
+	  go test -fuzz=FuzzValidateTransaction -fuzztime=2m ./evmcore/ ;\
+     cp -r ${SEEDDIR}* ${TEMPSEEDDIR} ;\
+     go test -v -run FuzzValidateTransaction -coverprofile=${GOCOVERDIR}/fuzz-validate-cover.out -coverpkg=${PACKAGES} ./evmcore/ ;\
+     go tool cover -html ${GOCOVERDIR}/fuzz-validate-cover.out -o ${GOCOVERDIR}/fuzz-validate-coverage.html ;\
+     rm -rf ${TEMPSEEDDIR} ;\
+
```
Comment (LuisPH3) [inline Makefile:78]: This may not be needed. 
The coverage report reflects the fuzzer function when used as an unit test. testing the hard-coded seeds only. 
I do not see how this coverage is more important than other fuzzers, to justify its own command in the Makefile. 

Let's avoid having too many make targets, it can get confusing very fast. (this target would need to be documented about what can it do and what not.... )

File: evmcore/tx_validation_fuzzer_test.go [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+
+// getTestTransactionsOptionFromRevision creates a validationOptions struct
+// with the specified revision and chain ID.
+func getTestTransactionsOptionFromRevision(revision int8, chainId *big.Int,
+	maxGas uint64, BaseFee, MinTip int64) validationOptions {
+	opt := validationOptions{
+		currentMaxGas:  maxGas,
+		currentBaseFee: big.NewInt(BaseFee),
+		minTip:         big.NewInt(MinTip),
+		isLocal:        true,
```
Comment (LuisPH3) [inline evmcore/tx_validation_fuzzer_test.go:489]:
```suggestion
		isLocal:        true,  /// localy submitted transactions have the more relaxed validation version. Therefore we test local true.
```

File: evmcore/tx_validation_fuzzer_test.go [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+		chainConfig,
+		vm.Config{},
+	)
+	evm.GasPrice = big.NewInt(int64(evmGasPrice))
+	return evm
+}
+
+// signTxForTest generates a new key, signs the transaction with it, and returns
+// the signer, address, and signed transaction.
+func signTxForTestWithChainId(t *testing.T, tx types.TxData, chainId *big.Int) (common.Address, *types.Transaction) {
```
Comment (LuisPH3) [inline evmcore/tx_validation_fuzzer_test.go:471]:
```suggestion
func signTxForTestWithChainId(t *testing.T, tx types.TxData, chainId *big.Int) (common.Address, *types.Transaction) {
    t.Helper()
```

**Review by LuisPH3 — CHANGES_REQUESTED** [review]

**Review by LuisPH3 — DISMISSED** [review]

File: Makefile [hunk 13]
```diff
@@ -61,6 +61,21 @@ integration-coverage:
...
+fuzz-validate-cover: SEEDDIR=$$(go env GOCACHE)/fuzz/github.com/0xsoniclabs/sonic/evmcore/FuzzValidateTransaction/
+fuzz-validate-cover: TEMPSEEDDIR=./evmcore/testdata/fuzz/FuzzValidateTransaction/
+fuzz-validate-cover:
+	@mkdir -p ${GOCOVERDIR} ;\
+     mkdir -p ${TEMPSEEDDIR} ;\
+	  go test -fuzz=FuzzValidateTransaction -fuzztime=2m ./evmcore/ ;\
+     cp -r ${SEEDDIR}* ${TEMPSEEDDIR} ;\
+     go test -v -run FuzzValidateTransaction -coverprofile=${GOCOVERDIR}/fuzz-validate-cover.out -coverpkg=${PACKAGES} ./evmcore/ ;\
+     go tool cover -html ${GOCOVERDIR}/fuzz-validate-cover.out -o ${GOCOVERDIR}/fuzz-validate-coverage.html ;\
+     rm -rf ${TEMPSEEDDIR} ;\
```
Comment (LuisPH3) [inline Makefile:77]: Please write an issue how to parametrize this target:
- by fuzzer function (this method can be used by all fuzzers) 
- by time, (2m is a very short time, we may be interested on much longer fuzzing times)

Comment (facuMH) [reply Makefile:77]: https://github.com/0xsoniclabs/sonic-admin/issues/178

File: evmcore/tx_validation_fuzzer_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go]: empty code shall be return before prague, for prague a non empty code is returned.

File: evmcore/tx_validation_fuzzer_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go]: these parameters are fuzzed over because it helped to keep the fuzzing simplified and allowed to test for more code paths.

File: evmcore/tx_validation_fuzzer_test.go [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+		// We intentionally avoid fuzzing these specific values to explicitly
+		// test the key edge cases: under, exact, and over.
+		// The magnitude of the difference doesn't matter;
+		// what's important is whether the value is below, equal to, or above
+		// the expected threshold.
+		for _, stateBalance := range []*big.Int{underCost, cost, overCost} {
+			for _, stateNonce := range []uint64{nonce - 1, nonce, nonce + 1} {
+				for _, validateGas := range []uint64{gas - 1, gas, gas + 1} {
+					for _, validateBaseFee := range []int64{feeCap - 1, feeCap, feeCap + 1} {
+						for _, validateMinTip := range []int64{tip - 1, tip, tip + 1} {
```
[the remark this answers is not in the document]
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go]: using `t.Run` was possible, but fuzzing over these parameters resulted in a faster execution of seeds and coverage was not significantly better for either options. So for simplicity this parameters are fuzzed over instead.

File: evmcore/tx_validation_fuzzer_test.go (hunk 12 above)
Comment (facuMH) [reply evmcore/tx_validation_fuzzer_test.go:471]: https://github.com/0xsoniclabs/sonic/pull/196/commits/3e5fc2282e58a8a126f03ac620fbc0804efd9c13

**Comment by LuisPH3** [comment]
This looks good, could you please run it over the long weekend?

**Review by LuisPH3 — DISMISSED** [review]

File: evmcore/tx_validation_fuzzer_test.go [hunk 15]
```diff
@@ -0,0 +1,517 @@
...
+		// affects intrinsic gas cost, these values are fuzzed,
+		// but the actual values are not relevant for this test.
+		accessListSize, authListSize uint,
+		// state
+		stateBalance, stateNonce, maxGas, baseFee, minTip uint64,
+	) {
+
+		if txType > types.SetCodeTxType {
+			// Skip invalid transaction types
+			return
```
Comment (HerbertJordan) [inline evmcore/tx_validation_fuzzer_test.go:203]: You may want to signal the skip explicitly.
```suggestion
			t.Skip()
```

File: evmcore/tx_validation_fuzzer_test.go [hunk 16]
```diff
@@ -0,0 +1,517 @@
...
+		}
+		if revision > testPrague {
+			// Skip invalid revisions
+			return
```
Comment (HerbertJordan) [inline evmcore/tx_validation_fuzzer_test.go:207]:
```suggestion
			t.Skip()
```

File: Makefile [hunk 17]
```diff
@@ -61,6 +61,21 @@ integration-coverage:
 integration-cover-all: PACKAGES=./...
 integration-cover-all: integration-coverage
 
+.PHONY: fuzz-validate-cover
```
Comment (HerbertJordan) [inline Makefile:64]: nit: for a global namespace like here in the make file, this name is a bit generic. It is unclear what "validate" is fuzzed.

Comment (facuMH) [reply Makefile]: there is an issue to parametrize this target (https://github.com/0xsoniclabs/sonic-admin/issues/178) there it should also be renamed to `make fuzz {testname} {fuzztime}`. 
But I will rename it now to `fuzz-txpool-validatetx-cover`

**Review by LuisPH3 — APPROVED** [review]
:+1:
